### PR TITLE
further prepare instancing

### PR DESCRIPTION
**Pull request by adRichter**

### CODE REVIEW

File: client/shader/plants.tcs [hunk 1]
```diff
@@ -101,6 +101,7 @@ void main() {
     }
 
     //handle tesselation level on distance
+
```
Comment (LukasKalbertodt) [inline client/shader/plants.tcs]: Please remove this empty line.

File: client/src/view/plant_view.rs [hunk 2]
```diff
@@ -127,15 +127,14 @@ fn gen_branch_buffer(old_cps: &[ControlPoint],
...
     for window in cps.windows(3) {
         let prev_cp = window[0];
         let curr_cp = window[1];
         let next_cp = window[2];
-
         let dir = prev_cp.point - next_cp.point;
 
-        for curr_point in &get_points_from_vector(dir) {
+        let tmp = get_points_from_vector(dir);
+        for curr_point in &tmp {
```
Comment (LukasKalbertodt) [inline client/src/view/plant_view.rs]: Noooo, someone edits my code :fearful: 

At least don't call it `tmp` :wink:

File: client/src/world/world_view.rs [hunk 3]
```diff
@@ -18,31 +19,24 @@ pub struct WorldView {
...
-        }
+        let chunks = HashMap::new();
+        let plant_views = HashMap::new();
 
         WorldView {
             chunks: chunks,
             chunk_renderer: chunk_renderer,
             plant_renderer: plant_renderer,
+            plant_views: plant_views,
+            plant_list: plant_list,
```
Comment (LukasKalbertodt) [inline client/src/world/world_view.rs]: You can just write `plant_views: HashMap::new()` ... we don't need to create the local variable `plant_views` for it (same goes for `chunks`). Less code is good in this case :)

File: client/src/world/world_view.rs [hunk 4]
```diff
@@ -56,10 +50,43 @@ impl WorldView {
...
                                                  facade));
+
+        for (pillar_pos, pillar) in chunk.pillars() {
+            for prop in pillar.props() {
+                // extract plant
+                let plant = &self.plant_list[prop.plant_index];
+
+                // match prop.prop {
+                //     PropType::Plant(ref x) => x,
+                // };
```
Comment (LukasKalbertodt) [inline client/src/world/world_view.rs]: Remove this please

File: client/src/world/world_view.rs [hunk 5]
```diff
@@ -56,10 +50,43 @@ impl WorldView {
...
+
+                let real_pos = pillar_pos.to_real();
+                let real_chunk_pos =
+                    AxialPoint::new(chunk_pos.0.q * (1 * world::CHUNK_SIZE as i32),
+                                    chunk_pos.0.r * (1 * world::CHUNK_SIZE as i32))
```
Comment (LukasKalbertodt) [inline client/src/world/world_view.rs]: I assume this multiplication with 1 is very important...

File: client/src/world/world_view.rs [hunk 6]
```diff
@@ -56,10 +50,43 @@ impl WorldView {
...
+                        .to_real();
+
+                if !self.plant_views.contains_key(&chunk_pos) {
+                    self.plant_views.insert(chunk_pos, Vec::new());
+                }
+                self.plant_views
+                    .get_mut(&chunk_pos)
+                    .unwrap()
```
Comment (LukasKalbertodt) [inline client/src/world/world_view.rs]: Have you seen the `entry` API? Beautifully designed -- with it you can remove 2 of those 3 map lookups:

```rust
self.plant_views.entry(chunk_pos).or_insert(Vec::new()).push( ... );
```

File: client/src/view/plant_view.rs (hunk 2 above)
Comment (adRichter) [inline client/src/view/plant_view.rs]: will be changed back was unnecessary

**Pull request merged**